Forum quickreply templatization

rica-carv opened the pull request:
### Motivation and Context
Enable custom template for the forum quick reply form

### Description
Changed code to take hardcoded HTML from forum quickreply shortcode to template

### How Has This Been Tested?
Latest e107 github code

### Types of Changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist
- [x] My code adheres to the e107 [code style standard](https://github.com/e107inc/e107/wiki/e107-Coding-Standard).
- [x] I have read the [**contributing** document](https://github.com/e107inc/e107/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/e107inc/e107/tree/master/e107_tests) to cover my changes.
- [ ] All new and existing tests pass.

rica-carv committed:
Update view_shortcodes.php

rica-carv committed:
Update view_shortcodes.php

rica-carv committed:
Update forum_viewtopic_template.php

CaMer0n commented:
@rica-carv Thank you for this! I left some comments. Just some minor tweaks and then I can commit it. Thank you.

rica-carv commented:
> I left some comments.

@CaMer0n Sorry, but i don't see any coments.... anywhere....

CaMer0n commented:
@rica-carv Sorry, It appears they went into a 'pending' state. Do you see them now?

rica-carv commented:
> Do you see them now?

@CaMer0n Yes i do. Tx

rica-carv commented on e107_plugins/forum/templates/forum_viewtopic_template.php:
```diff
@@ -481,6 +481,17 @@
 									";
 
+$FORUM_VIEWTOPIC_TEMPLATE['quickreply'] = "
+						<form action='{QR_URL}' method='post'>
```
Just by curiosity, why?

rica-carv replied:
@CaMer0n Done

rica-carv committed:
Update view_shortcodes.php

rica-carv committed:
Update forum_viewtopic_template.php

rica-carv commented on e107_plugins/forum/templates/forum_viewtopic_template.php:
```diff
@@ -481,6 +481,17 @@
+							<textarea cols='80' placeholder='{LAN=FORUM_2007}' rows='4' id='forum-quickreply-text' class='tbox input-xxlarge form-control' name='post' onselect='storeCaret(this);' onclick='storeCaret(this);' onkeyup='storeCaret(this);'></textarea>
+						</div>
+						<div class='center text-center form-group'>
+							<input type='submit' data-token='{QR_TOKEN}' data-forum-insert='{QR_AJAX}' data-forum-post='{QR_THFORUMID}' data-forum-thread='{QR_THID}' data-forum-action='quickreply' name='reply' value='{LAN=FORUM_2007}' class='btn btn-success button' />
```
@CaMer0n Done

rica-carv committed:
Update view_shortcodes.php

rica-carv commented:
Done.
Added default template for legacy purposes, althought it's akward to remove embbebed HTML asd then left it there for legacy purposes....

CaMer0n commented on e107_plugins/forum/shortcodes/batch/view_shortcodes.php:
```diff
@@ -1385,8 +1404,15 @@ function sc_quickreply()
+		return "<input type='submit' data-token='".e_TOKEN."' data-forum-insert='".(($thread->pages == $thread->page || $thread->pages == 0) ? 1 : 0)."' data-forum-post='".$this->var['thread_forum_id']."' data-forum-thread='".$this->var['thread_id']."' data-forum-action='quickreply' name='reply' value='".($parms['value']??LAN_FORUM_2007)."' class='btn btn-success button'>";
+	}
 
+	function sc_qr_hidden($parms = null)
```
You can place the hidden field inside the button shortcode.

rica-carv replied:
I don't think is good idea.
I've seen forms with the hidden fields in odd places inside the form.
This way we can have some flexibility... I think...

CaMer0n replied:
Indeed, it's better to place it before </form>.

rica-carv replied:
Force inject donne, in case it's missing from the designer template....

I think it's all covered now...

Let me know if you want any other change...

CaMer0n commented on e107_plugins/forum/templates/forum_viewtopic_template.php:
```diff
@@ -481,6 +481,14 @@
 
+$FORUM_VIEWTOPIC_TEMPLATE['quickreply'] = "
+						<div class='form-group'>
+							<textarea cols='80' placeholder='{LAN=FORUM_2007}' rows='4' id='forum-quickreply-text' class='tbox input-xxlarge form-control' name='post' onselect='storeCaret(this);' onclick='storeCaret(this);' onkeyup='storeCaret(this);'></textarea>
```
textarea should also be in a shortcode. The placeholder can be the default LAN, unless the shortcode receives a parm called 'label'

rica-carv replied:
Ok

CaMer0n commented on e107_plugins/forum/templates/forum_viewtopic_template.php:
```diff
@@ -481,6 +481,14 @@
+							<textarea cols='80' placeholder='{LAN=FORUM_2007}' rows='4' id='forum-quickreply-text' class='tbox input-xxlarge form-control' name='post' onselect='storeCaret(this);' onclick='storeCaret(this);' onkeyup='storeCaret(this);'></textarea>
+						</div>
+						<div class='center text-center form-group'>
+							{QR_SBUTTON:value=".LAN_FORUM_2007."}
```
Best to just make LAN_FORUM_2007 the default (within the shortcode) unless $parm['label'] is found.

CaMer0n commented:
Thanks. It's definitely better. Unfortunately, all the unit tests failed. Adding a defset('e_TOKEN') on e_TOKEN should fix it. 
I added some more notes. We want to separate functionality from design, so that a theme developer can't break the functionality of the form. 

 ---------
1) e_parse_shortcodeTest: Forum view shortcodes
 Test  tests/unit/e_parse_shortcodeTest.php:testForumViewShortcodes
                                        
Error: ;1m  [Error] Undefined constant "e_TOKEN"  
                                        
#1  /__w/e107/e107/e107_plugins/forum/shortcodes/batch/view_shortcodes.php:1411
#2  /__w/e107/e107/e107_handlers/shortcode_handler.php:1154
#3  e_parse_shortcode->doCode
#4  /__w/e107/e107/e107_handlers/shortcode_handler.php:1020
#5  /__w/e107/e107/e107_handlers/e_parse_class.php:854
#6  /__w/e107/e107/e107_tests/tests/unit/e_parse_shortcodeTest.php:1717
#7  /__w/e107/e107/e107_tests/tests/unit/e_parse_shortcodeTest.php:1367
#8  /__w/e107/e107/e107_tests/vendor/bin/codecept:119

rica-carv commented:
@CaMer0n I've got some more pull regarding moving HTML to templates, if you have the time to have a look...

rica-carv committed:
Update forum_viewtopic_template.php

rica-carv committed:
Update view_shortcodes.php

rica-carv committed:
Merge branch 'e107inc:master' into forrum_quickreply_templatization

rica-carv committed:
Update view_shortcodes.php

rica-carv committed:
Update view_shortcodes.php

rica-carv committed:
Merge branch 'e107inc:master' into forrum_quickreply_templatization

rica-carv commented:
@CaMer0n Think got it all donne... can you review it, please?

rica-carv committed:
Merge branch 'master' into forrum_quickreply_templatization

rica-carv committed:
Merge branch 'e107inc:master' into forrum_quickreply_templatization

rica-carv committed:
Merge branch 'e107inc:master' into forrum_quickreply_templatization

rica-carv committed:
Merge branch 'e107inc:master' into forrum_quickreply_templatization

rica-carv committed:
Merge branch 'e107inc:master' into forrum_quickreply_templatization

rica-carv committed:
Merge branch 'e107inc:master' into forrum_quickreply_templatization

e107help committed:
refactor(forum): finish the quick-reply extraction

Carries the design CaMer0n asked for in review to a state that can be read.

- `!strpos($template, "{QR_HIDDEN}")` treated a match at offset 0 as "not
  found" and appended a second hidden field. Compare against false.
- Read the block through `e107::getTemplate()` with `$merge = true` instead of
  the `$FORUM_VIEWTOPIC_TEMPLATE` global, so a theme that overrides
  forum_viewtopic_template.php without declaring 'quickreply' falls back to the
  core block rather than rendering an empty form.
- Restore LAN_FORUM_2006 as the submit label when the editor is enabled; the
  extraction had left LAN_FORUM_2007 on both paths.
- Guard `$thread` in the ajax-insert test, matching sc_nextprev() in this file.
- Remove the commented-out bodies the extraction left behind, and restore the
  unrelated whitespace edits to faqs_template.php and siteinfo_shortcodesTest.php.
